### PR TITLE
test: [M3 8035] - Refactor Marketplace App (OCA) Cypress test

### PR DESCRIPTION
**Pull request by dmcintyr-akamai**
## Description 📝

Refactor tests for one-click-apps to improve reliability. 

## Changes  🔄

List any change(s) relevant to the reviewer.

- Use random app from list of OCAs, instead of always using the first one
-"Fix" string input from OCA descriptions by replacing HTML special entities (such as "&amp;") with the html character ("&" in this case). Replacing only the ampersand and registered trademark entities for now, bc those are the only entities used in the descriptions in src/features/OneClickApps/oneClickApps.ts. 
 

## How to test 🧪
-yarn cy:run -s "packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts"

 

### Reproduction steps

Tests sporadically fail. The tests in the current develop branch will certainly fail if you hardcode the 2nd and 3rd tests in packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts to use a OCA description that has html entities (such as 595742 or 1132204 for an ID).  

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

### CODE REVIEW

File: packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts [hunk 1]
```diff
@@ -257,3 +254,39 @@ describe('OneClick Apps (OCA)', () => {
     ui.toast.assertMessage(`Your Linode ${linode.label} is being created.`);
   });
 });
+
+/**
+ * Returns the id of a randomly selected oneClickApp
+ * @returns number
+ */
+function getRandomApp(): number {
```
Comment (abailly-akamai) [inline packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:262]: Couple things here for improving code quality/DX (for both utils)
- we try to have our utils shared (& exported from) in `cypress/support/utils`
- alway good to be a bit more verbose with your function name, it helps with cognitive load while reading/refactoring code. ex: `getRandomOCAId`
- Update JDDoc comment to clarify we're returning an ID, not an OCA object

File: packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts [hunk 2]
```diff
@@ -257,3 +254,39 @@ describe('OneClick Apps (OCA)', () => {
...
+    '&reg;': '®',
+  };
+
+  const strDecoded = str.replace(/&[\w#]+;/g, (entity) => {
+    return htmlEntities[entity] || entity;
+  });
+  // replace any tab characters. some descriptions (ie, OCA id 595742) have double spaces that get converted to tabs
+  // also need to replace line breaks, either <br/> or <br> (see OCA id 1132204)
+  return strDecoded.replace('\t', '').replace(/<br ?\/?>/g, '');
+};
```
Comment (abailly-akamai) [inline packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:292]: Could we use `DOMParser` rather than a custom regex?

- It handles all HTML entities automatically, not just the ones we explicitly list
- It's maintained by the browser and will stay up to date with new entities
- It's likely more performant than using regex replacements

**Review by abailly-akamai — APPROVED** [review]
Thanks! PR looks good and refactor appears to solve the problem (at the very least locally) ✅ 

Approving pending consideration of the code improvements suggested

File: packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts (hunk 2 above)
Comment (bnussman-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:292]: We could just use `he` here, which is what the application code currently uses to parse the HTML entities, but I wasn't sure if the intention here was to ensure `he` is doing what it is suppose to be doing. In my opinion, we don't need to be testing third party libraries.

```ts
import { decode } from 'he';
```

DOMParser sounds neat. I don't know anything about it. It would be sweet if we could drop the `he` dependency for something native.

Comment (dmcintyr-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:292]: DOMParser would be better if i could get it to work equivalently to the code above, but the dom parser doesn't convert the entity for the ampersand. https://jsfiddle.net/mL57uvqb/

Comment (dmcintyr-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:292]: If this becomes a maintainability nightmare (or if you think it already is), i would just remove the problematic OCAs w/ the anomalous text (or else remove the special entities from the descriptions), bc the purpose of the test is to test the document structure and not the parsing. Right now my code for the edge case is just testing the cypress parser.

Comment (dmcintyr-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:292]: @bnussman-akamai the he.decode works better than DOMParser, but the test fails bc it encodes double spaces as " /t" (a tab after a space). The string in our test data accurately reflects the description for cPanel stored in our db. Afaik double spaces shouldn't be used and the browser just collapsed 2 spaces to one. And the cPanel description on other sites (https://azuremarketplace.microsoft.com/en-us/marketplace/apps/cpanel.cpanel?tab=overview) is similar but doesn't have double spaces. Imo it would be best just to fix this bad string w/: `decode(str.replace('  ', ' '));`

Comment (abailly-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:282]: As a general rule we should avoid installing a new package, esp if only serving an isolated case.

I think finding a workaround and/or shifting the testing strategy a little would be preferable if we're not giving up crucial coverage.

That being said, DOMParser can still do the job (your fiddle wasn't accurate). Try something like

```TS
const replaceHTMLEntities = (str: string) => {
  const parser = new DOMParser();
  return parser.parseFromString(`<!doctype html><body>${str}`, 'text/html').body.textContent || '';
};
```

Hopefully this can serve your purpose

Comment (bnussman-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:282]: For the record, I didn't install `he`, I was here long before I got here 😅 

Hopefully @abailly-akamai's example can work and we can drop `he`!

Comment (abailly-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:282]: ohhhh yeah that would be nice

Comment (dmcintyr-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:282]: > ```ts
> body.textContent
> ```

@abailly-akamai what result are you expecting to see from your snippet? i still don't get a "&" from your snippet. https://jsfiddle.net/mL57uvqb/4/

The DOMParser converts hex ("&#38;") or decimal format to html entities ("&amp;") but it doesn't convert "&amp;" to "&" which is what cypress does, so my expected string value never has the expected html character of "&".

File: packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts (hunk 1 above)
Comment (dmcintyr-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:263]: what would you suggest instead of "Returns the id of a randomly selected oneClickApp"?

Comment (abailly-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts:263]: my bad, that looks good!

File: packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts (hunk 2 above)
Comment (abailly-akamai) [reply packages/manager/cypress/e2e/core/oneClickApps/one-click-apps.spec.ts]: oh DOMParser won't work with Node... only in the browser

Screenshot 2025-02-04 at 13 37 23

never mind - sorry for the wrong suggestion

**Review by jdamore-linode — COMMENTED** [review]
Hey @dmcintyr-akamai, sorry I still haven't approved this -- I'm running into test failures for certain One-Click Apps. I was hoping I'd be able to find some solution I could share but it looks like I'll have to pick up where I left off tomorrow.

In the meantime, these are the app IDs I'm having trouble with: `401706`, `595742`, `607488`, `632758`, `1132204`.

One change that has made troubleshooting a bit easier is I've added a test attribute to [the element containing the app description](https://github.com/linode/manager/blob/develop/packages/manager/src/features/Linodes/LinodeCreate/Tabs/Marketplace/AppDetailDrawer.tsx#L137). Instead of using `cy.findByText`, I can now use `cy.get([data-qa-app-description]).should('have.text', '...')` which allows Cypress to tell us what content it's finding vs what it's expecting.

Some other things I've been exploring that might play a role in the solution:

- Whether or not we need to use [the `sanitizeHTML` util](https://github.com/linode/manager/blob/develop/packages/manager/src/utilities/sanitizeHTML.ts#L25), which we use in the app when rendering app descriptions
- Whether or not we can benefit from using the `should('have.html', '...')` assertion instead of `'have.text'`.

(For example, using `sanitizeHTML` _and_ `have.html` seems to work for the Redis app, which has `<br />` tags, but not the others)

This has become such a time sink over such a relatively minor detail that I'm nearly willing to remove the description assertion and give it another try later as a separate ticket, but I want to give it one more look tomorrow.

File: packages/manager/cypress/support/util/one-click-apps.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Returns the id of a randomly selected oneClickApp
+ * @returns number
+ */
+export function getRandomOCAId(): number {
+  // pick a random app
+  const appKeys = Object.keys(oneClickApps);
+  const index = Math.floor(Math.random() * appKeys.length);
+  // id should be number, so "+" useful to coerce from string
+  const id = +appKeys[index];
+  return id;
```
[the remark this answers is not in the document]
Comment (dmcintyr-akamai) [reply packages/manager/cypress/support/util/one-click-apps.ts:14]: for test purposes i shd probably test all of the apps instead of a random one or one that i know fails.

[the remark this answers is not in the document]
Comment (jdamore-linode) [reply packages/manager/cypress/support/util/one-click-apps.ts:14]: Yeah, I had the test running against all those IDs I mentioned before doing something like this:

```typescript
[401706, 595742, 607488, 632758, 1132204].forEach((candidateStackScriptId) => {
it('Can view app details of a marketplace app', () => {
  // Temporarily comment out the line where `candidateStackScript` is defined...
  ...
});
});
```

**Review by cliu-akamai — APPROVED** [review]
Tests look good to me.
